[lld] Refactor storage of PAuth ABI core info

sivan-shani opened the pull request:
Previously, the AArch64 PAuth ABI core values were stored as an ArrayRef<uint8_t>, introducing unnecessary indirection.

This patch replaces the ArrayRef with two explicit uint64_t fields: aarch64PauthAbiPlatform and aarch64PauthAbiVersion. This simplifies the representation and improves readability.

No functional change intended, aside from improved error messages.

sivan-shani committed:
[lld][GNU Properties] Refactor storage of PAuth ABI core info

Previously, the AArch64 PAuth ABI core values were stored as an ArrayRef<uint8_t>,
introducing unnecessary indirection.

This patch replaces the ArrayRef with two explicit uint64_t fields:
aarch64PauthAbiPlatform and aarch64PauthAbiVersion. This simplifies the
representation and improves readability.

No functional change intended.

sivan-shani commented:
This PR is a **preparatory refactor** to support a follow-up patch that will add **AArch64 Build Attributes** to LLD. The intent is to split the work into clean, reviewable steps and avoid repeating past mistakes.

Specifically, this change isolates and simplifies the handling of GNU Properties, so that support for Build Attributes can later be added **on top of a clearer structure**, with minimal duplication and maximum alignment between the two.

We would appreciate **as many eyes as possible** on this PR to ensure it is solid and aligns with project expectations, especially after feedback on the previous attempt (PR #131990). This is the first step in a more structured and minimal series of changes.

nasherm reviewed (commented):
A few comments but overall looking good

sivan-shani commented on lld/ELF/Config.h:
```diff
@@ -139,6 +140,11 @@ enum class GcsPolicy { Implicit, Never, Always };
 enum class ReportPolicy { None, Warning, Error };
 
+struct AArch64PauthAbiCoreInfo {
+  uint64_t aarch64PauthAbiPlatform;
```
The PAuth ABI spec defines both Platform Identifier and Version Number as 64-bit values:
https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#core-information

While the actual values may use fewer bits in practice, using uint64_t is necessary to match the specification.

sivan-shani commented on lld/ELF/Config.h:
```diff
@@ -695,7 +701,7 @@ struct Ctx : CommonLinkerContext {
   llvm::raw_fd_ostream openAuxiliaryFile(llvm::StringRef, std::error_code &);
 
-  ArrayRef<uint8_t> aarch64PauthAbiCoreInfo;
+  std::optional<AArch64PauthAbiCoreInfo> aarch64PauthAbiCoreInfo;
```
`std::optional` seems to fits here. This is a simple value-type struct that may or may not be present, and `optional` makes that clear without needing dynamic allocation or extra indirection.

A raw pointer would mean managing lifetime manually or implying ownership we don’t intend. With `optional`, it’s clear that the value lives directly in the parent object and is just conditionally present. It also avoids null checks and makes the intent more explicit.

ostannard reviewed (commented):
This isn't NFC, because it changes the error messages, but I think it's fine to do that since the new errors are more useful.

smithp35 reviewed (commented):
Main suggestion is to make aarch64PauthAbiPlatform a class as there is at least one invariant, and we can simplify some of the calling code by using the member functions. I'm not sure whether that will make a significant difference. I suggest trying it and then seeing if you prefer it.

I recommend taking out [GNU Properties] from the title as I don't think that has been used anywhere else and is not common enough to add any more information.

sivan-shani commented:
> This isn't NFC, because it changes the error messages, but I think it's fine to do that since the new errors are more useful.

Ack, commit message has been updated.

sivan-shani commented on lld/ELF/Config.h:
```diff
@@ -29,6 +29,7 @@
 #include "llvm/Support/GlobPattern.h"
 #include "llvm/Support/TarWriter.h"
 #include <atomic>
+#include <cstdint>
```
Seems as it is no needed, deleting.

sivan-shani commented:
Thanks everyone for the feedback.
To keep things manageable, I’ll address the comments in two commits:
The first will take care of the general suggestions—naming, cleanups, etc.
The second will focus on turning AArch64PauthAbiCoreInfo into a class, with constraints and some helper methods.
I hope that sounds reasonable.

sivan-shani committed:
Address general code review feedback

- Remove redundant include
- Rename fields in AArch64PauthAbiCoreInfo
- Simplify assignment using direct initialization

sivan-shani committed:
Add helper methods to AArch64PauthAbiCoreInfo

- Convert AArch64PauthAbiCoreInfo to a struct with isValid(), size(), and comparison operators.
- Simplify comparison logic in readSecurityNotes() using operator!=.

smithp35 reviewed (commented):
Just a request for a couple of small comment related changes and one case where we could use the size() function.

With that then I'm happy to move past the draft stage.

sivan-shani committed:
Address review feedback: clarify core info usage and use size() helper
- Added comments explaining the semantics of AArch64PauthAbiCoreInfo and the reserved (0, 0) value.
- Replaced hardcoded sizeof(uint64_t) * 2 with AArch64PauthAbiCoreInfo::size().

sivan-shani commented:
> With that then I'm happy to move past the draft stage.

Thanks, comments have been addressed. If no further feedback comes in, I plan to move this from 'Draft' to 'Ready for review' early next week.

sivan-shani commented:
The implementation of AArch64 Build Attributes on top of this change:
https://github.com/llvm/llvm-project/pull/142637

MaskRay reviewed (approved):
LGTM. Can an Arm reviewer confirm that the function is correct?

smithp35 reviewed (approved):
I can confirm that this is functionally correct. The only differences are in the error message when there's a clash.

sivan-shani commented:
Thank you, merging.
